internal/connection/connection_pool.go

```diff
@@ -49,6 +51,19 @@ func (cp *ConnectionPool) Put(key string, conn *Connection) {
 	cp.pool[key] = conn
 }
 
+// Get returns the connection object corresponding to given key.
+func (cp *ConnectionPool) Get(key string) *Connection {
+	cp.rwlock.Lock()
+	defer cp.rwlock.Unlock()
```

Comment:
```suggestion
	cp.rwlock.RLock()
	defer cp.rwlock.Runlock()
```

I guess just read lock is enough here.